kratos/python_interface/python_registry.py

```diff
@@ -0,0 +1,372 @@
+import KratosMultiphysics
+from KratosMultiphysics.kratos_utilities import GetListOfAvailableApplications
+
+class RegistryContext():
+    ALL = 0
+    CPP = 1
+    PYTHON = 2
+
+class PythonRegistryIterator:
+    ''' PythonRegistry iterator class
+    '''
+    def __init__(self, ThisPythonRegistry):
+       # Member variables to keep track of iteration
+       self.__index = 0
+       self.__cpp_root_keys = ThisPythonRegistry._PythonRegistry__cpp_registry.keys()
+       self.__py_root_keys = list(ThisPythonRegistry._PythonRegistry__python_registry.keys())
+
+    def __next__(self):
+        '''Returns the next object from the PythonRegistry wrapper root level
+        '''
+        n_py_items = len(self.__py_root_keys)
+        n_total_items = len(self.__cpp_root_keys) + n_py_items
+        if self.__index < n_total_items:
+            # Check if we are at the Python or c++ registry
+            if self.__index < n_py_items:
+                # Return the item key from the Python registry
+                item_key = self.__py_root_keys[self.__index]
+            else:
+                # Return the item key from the c++ registry
+                item_key = self.__cpp_root_keys[self.__index-n_py_items]
+            self.__index += 1
+            return item_key
+        # End of iteration
+        raise StopIteration
+
+class PythonRegistry(object):
+
+    #TODO: Change this by a frozenmap() once this is available
+    CppGetFunctionNamesMap = {
+        "Operations": "GetOperation",
+        "Processes": "GetProcess"
+    }
+
+    def __init__(self):
+        # We get a "private" pointer to the c++ registry
+        self.__cpp_registry = KratosMultiphysics.CppRegistry
```

Comment:
This is a mechanism to have an alternative access to the `CppRegistry`. After doing this one could still access the `CppRegistry` from python by doing `KratosMultiphysics.CppRegistry.WatheverCppFunction()`. This is something we want to be explicitly forbidden in order to force people to access the c++ registry through the Python extension, implying that, the user doesn't even realize that there is a c++ layer and a Python extension one but a unique registry.

Besides, we do the `__getattr__` override at the module level in order to throw a meaningful exception if the user still calls `KratosMultiphysics.CppRegistry`.

Once we've saved this pointer, we do the `pop` in your comment above in order to explicitly remove the `KratosMultiphysics.CppRegistry` access so we only keep the `self__cpp_registry` inside the Python layer. We need to do this `pop` instead of using pybind properties as we do in normal classes because the `CppRegistry` is indeed a static global variable. If we were not doing this `pop` the exception would never be thrown.

I tried to be as clear as possible but maybe it's still a bit blurry. Maybe @roigcarlo can give a better explanation.

Reply:
Got it, thanks!

Reply:
Resolving!